### PR TITLE
Serial: better exit

### PR DESCRIPTION
**Pull request by khancyr**
Hello,

This should prevent the "Error: serial0: receive: End of file" problem !

### CODE REVIEW

File: libmavconn/src/serial.cpp [hunk 1]
```diff
@@ -54,6 +54,15 @@ MAVConnSerial::MAVConnSerial(uint8_t system_id, uint8_t component_id,
...
 		serial_dev.set_option(SPB::flow_control( (hwflow) ? SPB::flow_control::hardware : SPB::flow_control::none));
+		#ifdef __linux__
+		int fd = serial_dev.native_handle();
+		struct termios tio;
+		tcgetattr(fd, &tio);
+		tio.c_cc[VTIME] = 0;
+		tio.c_cc[VMIN] = 1;
+		tcflush(fd, TCIFLUSH);
+		tcsetattr(fd, TCSANOW, &tio);
+		#endif
```
Comment (vooon) [inline libmavconn/src/serial.cpp]: It is possible to do that via ASIO API? Just `set_option()` looks a bit better (but it also does tty calls).

Ifdef should not have any indentation. And maybe enclosure that hack to block?
Also it definitely should have comment on what problem solves that hack.

**Comment by TSC21** [comment]
@khancyr any updates here?

**Comment by khancyr** [comment]
I will do it before the end of the week, tcp and udp would gains from this too

**Comment by khancyr** [comment]
@vooon set_option() doesn't allow to set those option. I have change to set to the serial port to raw mode that should prevent the EOF error.
The other commit is just to ensure that everything is close cleanly.

File: libmavconn/src/serial.cpp [hunk 2]
```diff
@@ -82,12 +90,16 @@ void MAVConnSerial::close()
...
-	io_service.stop();
+	serial_dev.cancel();
 	serial_dev.close();
 
+	io_service.stop();
+
 	if (io_thread.joinable())
 		io_thread.join();
 
+	io_service.reset();
```
Comment (vooon) [inline libmavconn/src/serial.cpp]: That is not necessary, destructor should free automatically.

File: libmavconn/src/tcp.cpp [hunk 3]
```diff
@@ -131,14 +131,18 @@ void MAVConnTCPClient::close()
 	lock_guard lock(mutex);
 	if (!is_open())
 		return;
+	socket.shutdown(boost::asio::ip::tcp::socket::shutdown_send);
```
Comment (vooon) [inline libmavconn/src/tcp.cpp]: Blank line before lost.

**Comment by vooon** [comment]
Thank you! I think i do release with that fix.

File: libmavconn/src/serial.cpp (hunk 2 above)
Comment (khancyr) [reply libmavconn/src/serial.cpp]: Normally yes, but that way we ensure that it is really done ! Using serial on asio should also put the port on raw mode but isn't done properly ;-P see https://github.com/chriskohlhoff/asio/commit/ff9d69e2fc002f62828d1c23b5eaae993427bc07

**Comment by mhkabir** [comment]
@vooon Please hold for release. I have another hotfix for flow control incoming. This is fixed in Boost >= 1.66, but we cannot have that as default version yet till ROS supports it.

**Pull request merged**